test/controllers/tasks_controller_test.rb

```diff
@@ -79,48 +82,122 @@
-    #        thing to test.
-    it "can update an existing task" do
-      # Your code here
+    before do 
+      Task.create(name:"Chores", description:"Mop the floor")
+
+      @new_task_hash = {
+        task: {
+        name: "Homework",
+        description: "Complete personal portfolio",
+        completed: DateTime.new(10/6/19),
+        },
+    }
+    end 
+
+    it "can update an existing task successfully and redirects to home" do
+      existing_task = Task.first 
+      id = existing_task.id
+
+      expect {
+        patch task_path(id), params: @new_task_hash 
+      }.wont_change "Task.count"
+
+      expect(Task.find_by(id: existing_task.id).description).must_equal "Complete personal portfolio"
+
+      expect(Task.find_by(id: existing_task.id).name).must_equal "Homework"
+
+      #Validating that the completed field holds a DateTime object, which I think is called this weird thing in Rails?
+      expect(Task.find_by(id: existing_task.id).completed).must_be_kind_of ActiveSupport::TimeWithZone
+
+      expect(Task.find_by(id: existing_task.id).completed).must_equal DateTime.new(10/6/19)
+
+      must_redirect_to task_path(existing_task.id)
+
     end
 
     it "will redirect to the root page if given an invalid id" do
-      # Your code here
+      invalid_id = [-1, "woops"]
+      invalid_id.each do |id|
+        patch task_path(invalid_id)
+        must_redirect_to root_path
+      end
     end
+
+    it "won't update a task with invalid form entries" do
+      expect { patch task_path(task.id), params: invalid_task_hash }.wont_change "task.updated_at"
+    end 
+
   end
 
-  # Complete these tests for Wave 4
   describe "destroy" do
-    # Your tests go here
+    it "destroys a task correctly, and redirect to root path afterwards" do 
+      
+      before_deletion_count = Task.count
+      delete task_path(task.id)
+      after_deletion_count = Task.count
+
+      assert (before_deletion_count == after_deletion_count)
+
+      must_redirect_to root_path
+    end 
 
+    it "redirects to task index and does not actually delete a nonexistent book" do
+      Task.destroy_all
+      invalid_task_id = 1
+
+      expect { delete task_path(invalid_task_id).must_differ "Task.count", 0 }
+    end 
+
+    it "redirects to task index if trying to delete a task that has already been deleted" do 
+      Task.create(name:"test", description:"TEST")
+      task_id = Task.find_by(name:"test").id
+      Task.destroy_all
+
+      expect{ delete task_path(task_id)}.must_differ "Task.count", 0
+
+      must_redirect_to root_path 
+    end 
   end
 
   # Complete for Wave 4
-  describe "toggle_complete" do
-    # Your tests go here
+  describe "completed task" do
+    it "completed changes from nil to current datetime, and redirects to home" do
+      #arrange
+      test_task = Task.create(name: "Testing", description: "in process")
+
+      #act
+      patch completed_task(id:test_task.id), params: {destination: "root"} 
```

Comment:
In order to get this to work, you'll need to do `put completed_tasks_path(id:test_task.id), params: {destination: "root"}`